README.md
This package is yet another Rust like Result implementation. 
There are already several others and they are basically all the same.
This isn't much different. Just pick one you like.

## Features

The Features include the basic functions you would expect:

- expect 
- unwrap
- unwrapOr
- unwrapOrElse
- andThen
- orElse

These functions also exist as extension Functions for Futures of Results for Convenience reasons.

One thing that is special about this implementation, is that you can use the andThen
and orElse functions to convert a result to a slightly different one. 
This can be usefull wenn you have multiple function calls in one function that return 
different Results. For more details on that have a look at the examples.

## Getting started

```dart
import "package:only_results/only_results.dart";

var result = Result.catchErrors(() => unsafeFunction());

switch(result) {
    case Ok(:void value): 
        print("it worked =)");
    case Err(:Object error): 
        print("=( $error");
}

```

## Usage

To use this library you just have to import it into your project and define the return type of your 
functions according to your needs.

```dart

void main() {
    var a = 42;
    var b = 10;
    switch(divide(a, b)) {
        case Ok(:int value): 
            print("$a / $b = $value");
        case Err(:String error):
            print("ERROR: $error");
    }
}

Result<int, String> divide(int a, int b) {
    if(b == 0) {
        return Err("divison by zero is not allowed");
    }
    return Ok(a ~/ b);
}
```

### Just handle the Happy Path

If you have for example written a function, that you know will, given your specific
input, always return an Ok of something, you can use either unwrap or expect.

```dart
void main() {
    var result = someFunction("Hallo Welt");

    int value = result.unwrap();
    int value2 = result.expect("Exception Text if result is an Error")
    int value3 = result.unwrapOr(42);
}

Result<int, String> someFunction(String value) {
    ...
}
```

### Handle the Result with andThen / orElse

As you have already seen you can use a switch-statement to handle the Result object.
But you can also use the andThen and orElse function to chain multiple actions 
together.

```dart
void main() {
    fn1("Hallo Welt").andThen(fn2).orElse(errHandler);
    fn1("Hallo Welt").andThen((okValue) => Ok(okValue / 2));
}

Result<int, String> fn1(String str) {
    ...
}

Result<int, String> fn2(int i) {
    ...
}

Result<int, String> errHandler(String str) {
    ...
}
```

You can also use those functions to convert a result to another type of result.

```dart
Result<int, String> findFirst(List<String> words, String word) {
  return Result.catchErrors(() => words.firstWhere((it) => it == word))
    .orElse((Object error) => Err(error.toString()));
}
```

## Additional information

If you want to see more examples, have a look at the tests.


CHANGELOG.md
## 1.0.0

- Initial version.

## 1.0.1

- updated platform support specification

## 1.0.2

- added implementation of hashCode getter and == operator
- added implementation for unwrapOrElse 
- added implementation for unwrapOrElseAsync


test/extensions_test.dart
import 'package:only_results/only_results.dart';
import 'package:test/test.dart';

late Future<Result<int, String>> ok;
late Future<Result<int, String>> err;

void main() {
  group("test Future Results", extensionTests);
}

void extensionTests() {
  setUp(setTestsUp);
  test("test Future expect", testExpect);
  test("test Future unwrap", testUnwrap);
  test("test Future unwrapOr", testUnwrapOr);
  test("test Future andThen", testAndThen);
  test("test Future andThenAsync", testAndThenAsync);
  test("test Future orElse", testOrElse);
  test("test Future orElseAsync", testOrElseAsync);
}

void setTestsUp() {
  ok = Future.value(Ok(42));
  err = Future.value(Err("fooo"));
}

Future<void> testExpect() async {
  expect(await ok.expect("bar"), 42);

  late String exceptionString;
  try {
    await err.expect("bar");
  } catch (e) {
    exceptionString = e.toString();
  }
  expect(exceptionString, "Exception: bar");
}

Future<void> testUnwrap() async {
  expect(await ok.unwrap(), 42);

  late String exceptionString;
  try {
    await err.unwrap();
  } catch (e) {
    exceptionString = e.toString();
  }

  expect(exceptionString, "Exception: fooo");
}

Future<void> testUnwrapOr() async {
  expect(await ok.unwrapOr(21), 42);
  expect(await err.unwrapOr(21), 21);
}

Future<void> testAndThen() async {
  expect(await ok.andThen((ok) => Ok(ok.toString())).unwrap(), "42");
  expect((await err.andThen((ok) => Ok(ok.toString())) as Err).error, "fooo");
}

Future<void> testAndThenAsync() async {
  expect(await ok.andThenAsync((ok) async => Ok(ok.toString())).unwrap(), "42");
  expect((await err.andThenAsync((ok) async => Ok(ok.toString())) as Err).error, "fooo");
}

Future<void> testOrElse() async {
  expect(
    await ok.andThen((ok) => Ok(ok.toString())).orElse((err) => Ok(err)).unwrap(),
    "42",
  );
  expect(
    await err.andThen((ok) => Ok(ok.toString())).orElse((err) => Ok(err)).unwrap(),
    "fooo",
  );
}

Future<void> testOrElseAsync() async {
  expect(
    await ok.andThen((ok) => Ok(ok.toString())).orElseAsync((err) async => Ok(err)).unwrap(),
    "42",
  );
  expect(
    await err.andThen((ok) => Ok(ok.toString())).orElseAsync((err) async => Ok(err)).unwrap(),
    "fooo",
  );
}


test/only_results_test.dart
import 'package:only_results/only_results.dart';
import 'package:test/test.dart';

late Result<int, String> ok;
late Result<int, String> err;

void main() {
  group("test Results", () {
    setUp(setTestsUp);
    test("test Result.catchErrors", testCatchErrorsOnStateError);
    test("test Ok.expect", testExpectOk);
    test("test Err.expect", testExpectErr);
    test("test Ok.unwrap", testUnwrapOk);
    test("test Err.unwrap", testUnwrapErr);
    test("test Result.isOk", testIsOk);
    test("test Result.isErr", testIsErr);
    test("test Ok.unwrapOr", testUnwrapOrOk);
    test("test Err.unwrapOr", testUnwrapOrErr);
    test("test Ok.unwrapOrElse", testUnwrapOrElseOk);
    test("test Err.unwrapOrElse", testUnwrapOrElseErr);
    test("test Ok.unwrapOrElseAsync", testUnwrapOrElseAsyncOk);
    test("test Err.unwrapOrElseAsync", testUnwrapOrElseAsyncErr);
    test("test Result.andThen", testAndThen);
    test("test Result.orElse", testOrElse);
    test("test Result.==", testEquals);
  });
}

void setTestsUp() {
  ok = Ok(42);
  err = Err("Foo");
}

void testCatchErrorsOnStateError() {
  var result = Result.catchErrors(() => [1, 2, 3].firstWhere((it) => it == 2));
  expect(result.isOk(), true);
  expect(result.isErr(), false);
  expect(result.unwrap(), 2);

  result = Result.catchErrors(() => <int>[].firstWhere((it) => it == 42));

  expect(result.isErr(), true);
  expect(result.runtimeType.toString(), "Err<int, Object>");
  expect((result as Err<int, Object>).error.runtimeType.toString(), "StateError");
}

void testExpectOk() {
  expect(ok.expect("irrelevant string"), 42);
}

void testExpectErr() {
  late Exception exception;
  try {
    err.expect("ALARM!!!");
  } catch (e) {
    exception = e as Exception;
  }
  expect(exception.toString(), "Exception: ALARM!!!");
}

void testUnwrapOk() {
  expect(ok.unwrap(), 42);
}

void testUnwrapErr() {
  late Exception exception;
  try {
    err.unwrap();
  } catch (e) {
    exception = e as Exception;
  }
  expect(exception.toString(), "Exception: Foo");
}

void testIsOk() {
  expect(Ok(42).isOk(), true);
  expect(Err("nope").isOk(), false);
}

void testIsErr() {
  expect(Ok(42).isErr(), false);
  expect(Err("nope").isErr(), true);
}

void testUnwrapOrOk() {
  expect(ok.unwrapOr(21), 42);
}

void testUnwrapOrErr() {
  expect(err.unwrapOr(21), 21);
}

void testUnwrapOrElseOk() {
  expect(ok.unwrapOrElse(() => 21), 42);
}

void testUnwrapOrElseErr() {
  expect(err.unwrapOrElse(() => 21), 21);
}

void testUnwrapOrElseAsyncOk() async {
  final result = await ok.unwrapOrElseAsync(() async => 21);
  expect(result, 42);
}

void testUnwrapOrElseAsyncErr() async {
  final result = await err.unwrapOrElseAsync(() async => 21);
  expectLater(result, 21);
}

void testAndThen() {
  var result = ok.andThen((ok) => Ok(ok * 2));

  expect(result.unwrap(), 84);
  expect(result.runtimeType, Ok<int, String>);

  var result2 = ok.andThen((ok) => Ok(ok.toDouble()));

  expect(result2.unwrap(), 42.0);
  expect(result2.runtimeType, Ok<double, String>);

  var result3 = Err<int, String>("Foo");

  expect(result3.unwrapOr(42), 42);
  expect(result3.runtimeType, Err<int, String>);
}

void testOrElse() {
  var result = err.orElse((err) => Ok(42));

  expect(result.unwrap(), 42);
  expect(result.runtimeType, Ok<int, dynamic>);

  var result2 = err.orElse((err) => Err(21));

  expect((result2 as Err<int, int>).error, 21);
  expect(result2.runtimeType, Err<int, int>);

  var result3 = Ok<int, String>(42).orElse((err) => Ok(21));
  expect(result3.unwrap(), 42);
}

void testEquals() {
  expect(Ok<String, String>("yes") == Ok<String, String>("yes"), true);
  expect(Ok<String, String>("yes") == Ok<String, String>("no"), false);
  expect(Ok<String, String>("yes") == Ok<String, int>("yes"), false);

  expect(Err<String, String>("yes") == Err<String, String>("yes"), true);
  expect(Err<String, String>("yes") == Err<String, String>("no"), false);
  expect(Err<String, String>("yes") == Err<int, String>("yes"), false);

  expect(Ok<String, String>("yes") == Err<String, String>("yes"), false);
  expect(Ok<String, String>("yes") == Err<String, String>("no"), false);
  expect(Ok<String, String>("yes") == Err<int, String>("yes"), false);
}


lib/src/extensions.dart
import 'only_results_base.dart';

///This extensions makes it easier to handle the [Result] inside
///of [Future] objects.
extension FutureExtensions<O, E> on Future<Result<O, E>> {
  /// awaits the [Future] and executes the expect of the [Result].
  Future<O> expect(String msg) async => (await this).expect(msg);

  /// awaits the [Future] and executes the unwrap of the [Result].
  Future<O> unwrap() async => (await this).unwrap();

  /// awaits the [Future] and executes the unwrapOr of the [Result].
  Future<O> unwrapOr(O defaultValue) async => (await this).unwrapOr(defaultValue);

  /// awaits the [Future] and executes the unwrapOrElse of the [Result].
  Future<O> unwrapOrElse(O Function() fn) async => (await this).unwrapOrElse(fn);

  /// awaits the [Future] and executes the unwrapOrElseAsync of the [Result].
  Future<O> unwrapOrElseAsync(Future<O> Function() fn) async =>
      await (await this).unwrapOrElseAsync(fn);

  /// awaits the [Future] and executes the andThen of the [Result].
  Future<Result<O2, E>> andThen<O2>(Result<O2, E> Function(O ok) fn) async {
    return (await this).andThen(fn);
  }

  /// awaits the [Future] and executes the andThenAsync of the [Result].
  Future<Result<O2, E>> andThenAsync<O2>(
    Future<Result<O2, E>> Function(O ok) fn,
  ) async {
    return await (await this).andThenAsync(fn);
  }

  /// awaits the [Future] and executes the orElse of the [Result].
  Future<Result<O, E2>> orElse<E2>(Result<O, E2> Function(E err) fn) async {
    return (await this).orElse(fn);
  }

  /// awaits the [Future] and executes the orElse of the [Result].
  Future<Result<O, E2>> orElseAsync<E2>(
    Future<Result<O, E2>> Function(E err) fn,
  ) async {
    return await (await this).orElseAsync(fn);
  }
}


lib/src/only_results_base.dart
/// [Result] is meant to handle errors in a way that makes the possibility of
/// errors visible by returning a [Result]. It also provides methods to handle
/// it. This library is an incomplete dart implementation of rusts results.
sealed class Result<O, E> {
  /// Wraps the return value of [fn] into an [Ok]. If fn throws an [Exception] or [Error] it is
  /// wrapped into an [Err].
  static Result<O, Object> catchErrors<O>(O Function() fn) {
    try {
      return Ok(fn());
    } catch (e) {
      return Err(e);
    }
  }

  /// Wraps the return value of [fn] into an [Ok]. If fn throws an [Exception] or [Error] it is
  /// wrapped into an [Err].
  static Future<Result<O, Object>> catchErrorsAsync<O>(
    Future<O> Function() fn,
  ) async {
    try {
      return Ok(await fn());
    } catch (e) {
      return Err(e);
    }
  }

  bool isErr();
  bool isOk();

  /// Returns the value if this [Result] is an [Ok]. If it is an [Err] it throws
  /// with the [msg] as message.
  O expect(String msg);

  /// Returns the value of an [Ok] object. If it is an [Err] object intead, it throws
  /// an Exception with the error as the message that is wrapped by the [Err].
  O unwrap();

  /// Returns the value of an [Ok] object. If it is an [Err] object, the defaultValue
  /// is returned instead.
  O unwrapOr(O defaultValue);

  /// Returns the value of an [Ok] object. If it is an [Err] object, it returns
  /// an alternate value calculated by [fn].
  O unwrapOrElse(O Function() fn);

  /// Returns the value of an [Ok] object wrapped in a [Future]. If it is an [Err]
  /// object, it returns an alternate value calculated asynchronously by [fn]
  /// wrapped in a [Future].
  Future<O> unwrapOrElseAsync(Future<O> Function() fn);

  /// If this [Result] is of type [Ok] then [fn] passed into this function is executed. It
  /// itself needs to return a [Result].
  /// if this [Result] is of typhan it is returned immediately.
  Result<O2, E> andThen<O2>(Result<O2, E> Function(O ok) fn);

  /// If this [Result] is of type [Ok] then [fn] passed into this function is executed. It
  /// itself needs to return a [Result].
  /// if this [Result] is of type [Err] than it is returned immediately.
  Future<Result<O2, E>> andThenAsync<O2>(
    Future<Result<O2, E>> Function(O ok) fn,
  );

  /// If this [Result] is of type [Ok] then it is returned immediatly.
  /// Otherwise [fn] is executed and it's [Result] is returned.
  Result<O, E2> orElse<E2>(Result<O, E2> Function(E err) fn);

  /// If this [Result] is of type [Ok] then it is returned immediatly.
  /// Otherwise [fn] is executed and it's [Result] is returned.
  Future<Result<O, E2>> orElseAsync<E2>(
    Future<Result<O, E2>> Function(E err) fn,
  );
}

class Ok<O, E> extends Result<O, E> {
  O value;

  Ok(this.value);

  @override
  bool isErr() => false;

  @override
  bool isOk() => true;

  @override
  O expect(String msg) => value;

  @override
  O unwrap() => value;

  @override
  O unwrapOr(O defaultValue) => value;

  @override
  O unwrapOrElse(O Function() fn) => value;

  @override
  Future<O> unwrapOrElseAsync(Future<O> Function() fn) async => value;

  @override
  Result<O2, E> andThen<O2>(Result<O2, E> Function(O ok) fn) => fn(value);

  @override
  Future<Result<O2, E>> andThenAsync<O2>(
    Future<Result<O2, E>> Function(O ok) fn,
  ) async {
    return await fn(value);
  }

  @override
  Result<O, E2> orElse<E2>(Result<O, E2> Function(E err) fn) => Ok(value);

  @override
  Future<Result<O, E2>> orElseAsync<E2>(
    Future<Result<O, E2>> Function(E err) fn,
  ) async {
    return Ok(value);
  }

  @override
  int get hashCode {
    return runtimeType.hashCode * 11 + value.hashCode;
  }

  @override
  bool operator ==(Object other) {
    return other is Ok<O, E> && other.value == value;
  }

  @override
  String toString() => "$runtimeType($value)";
}

class Err<O, E> extends Result<O, E> {
  E error;

  Err(this.error);

  @override
  bool isErr() => true;

  @override
  bool isOk() => false;

  @override
  O expect(String msg) => throw Exception(msg);

  @override
  O unwrap() => throw Exception(error);

  @override
  O unwrapOr(O defaultValue) => defaultValue;

  @override
  O unwrapOrElse(O Function() fn) => fn();

  @override
  Future<O> unwrapOrElseAsync(Future<O> Function() fn) async => await fn();

  @override
  Result<O2, E> andThen<O2>(Result<O2, E> Function(O ok) fn) => Err(error);

  @override
  Future<Result<O2, E>> andThenAsync<O2>(
    Future<Result<O2, E>> Function(O ok) fn,
  ) async {
    return Err(error);
  }

  @override
  Result<O, E2> orElse<E2>(Result<O, E2> Function(E err) fn) => fn(error);

  @override
  Future<Result<O, E2>> orElseAsync<E2>(
    Future<Result<O, E2>> Function(E err) fn,
  ) async {
    return await fn(error);
  }

  @override
  int get hashCode {
    return runtimeType.hashCode * 11 + error.hashCode;
  }

  @override
  bool operator ==(Object other) {
    return other is Err<O, E> && other.error == error;
  }

  @override
  String toString() => "$runtimeType($error)";
}


example/only_results_example.dart
import 'package:only_results/only_results.dart';

void main() {
  var result = Result.catchErrors(() => <int>[].first);

  switch(result) {
      case Ok(:int value): 
          print("$value: it worked =)");
      case Err(:Object error): 
          print("=( $error");
  }
}

//////////////////////////////////////////////////////////
// Simple handling of errors
//////////////////////////////////////////////////////////
void simpleErrorHandlingExample() {
    var a = 42;
    var b = 10;
    switch(divide(a, b)) {
        case Ok(:int value): 
            print("$a / $b = $value");
        case Err(:String error):
            print("ERROR: $error");
    }
}

Result<int, String> divide(int a, int b) {
    if(b == 0) {
        return Err("divison by zero is not allowed");
    }
    return Ok(a ~/ b);
}

//////////////////////////////////////////////////////////
// Unwraping results
//////////////////////////////////////////////////////////

void unwrapIt() {
  late Result<int, String> result;
  result = Ok(42);

  //value is 42
  var value = result.unwrap();
  
  //value is 21
  value = result.unwrapOr(21);

  //value is 42
  value = result.expect("Error Text");

  result = Err("ALARM");

  try {
    value = result.unwrap();
  } catch (e) {
    // prints "Excpetion: ALARM"
    print(e);
  }

  //value is 21
  value = result.unwrapOr(21);

  try {
    value = result.expect("Error Text");
  } catch (e) {
    // prints "Exception: Error Text"
    print(e);
  }
}

//////////////////////////////////////////////////////////
// handling with andThen/OrElse
//////////////////////////////////////////////////////////

void andThenOrElse() {
  // results in Ok<int, String>(84)
  Ok<int, String>(42).andThen((ok) => Ok(ok*2)).orElse((err) => Err("Foo: $err"));
  // results in Err<int, String>("Foo: bar")
  Err<int, String>("bar").andThen((ok) => Ok(ok*2)).orElse((err) => Err("Foo: $err"));

  // results in Ok<double, int?>
  Ok<int, String>(42).andThen((ok) => Ok(ok.toString())).orElse((err) => Err(int.tryParse(err)));
  // results in Err<double, int?>(42)
  Err<int, String>("42").andThen((ok) => Ok(ok.toString())).orElse((err) => Err(int.tryParse(err)));
}
